### ISSUE TITLE
Storage I/O tracing

### ISSUE THREAD

**Issue by danmih-ixia** [mlcommons/chakra#222]
## Problem Related to the Feature
> Back in PR [#216](https://github.com/mlcommons/chakra/pull/216) we landed the typed storage schema — `storage.proto`, `STORAGE_NODE`, `DATA_OP_NODE`, the `op_info` oneof.
> The schema is in `main`, but nothing produces it and nothing consumes it yet.

## Proposed Solution
> We'd like to propose a tracing solution for capturing storage I/O from a PyTorch training run into a Chakra execution trace.

# kaitrace — architecture and tracing workflow

How kaitrace captures storage I/O from a PyTorch training run and turns it into a
Chakra execution trace with typed `STORAGE_NODE` / `DATA_OP_NODE` and `StorageInfo`.

---

## 1. Component architecture

Four layers plus the converter pass. **Only layer 3 needs root** — layers 1, 2
and 4 are pure Python and run anywhere torch does.

```mermaid
flowchart TB
    subgraph L1["Layer 1 · Torch-side capture — pure Python, no root, portable"]
        direction LR
        EN["<b>enable.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>enable_storage_capture — installs everything</li><li>registers the dataset's block devices with the sidecar</li></ul>"]
        PF["<b>posix_fallback.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>open / read / write / close</li><li>emits storage_op fs regions</li><li>registers dataset file inodes at open</li></ul>"]
        DT["<b>dataloader_tracker.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>DataLoader lifecycle</li><li>recipe / iter / consume markers</li><li>make_worker_init_fn — a per-worker ExecutionTraceObserver</li><li>patches _worker_loop to flush before os._exit</li></ul>"]
        DW["<b>dataset_wrapper.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>wraps Dataset.__getitem__</li><li>stamps idx and worker_id</li></ul>"]
        BS["<b>batch_sync.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>worker to main handoff</li><li>produce / consume pairing</li></ul>"]
        TS["<b>torch_save_patch.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>torch.save / torch.load</li><li>emits checkpoint regions</li></ul>"]
        EN --> PF & DT & BS & TS
        DT --> DW
    end

    subgraph L2["Layer 2 · The contract — the whole producer/consumer ABI"]
        direction TB
        PAT["<b>_patterns.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>encode kind + fields to a pipe-joined name</li><li>kinds: storage_op · batch · checkpoint · chakra_global_metadata</li></ul>"]
        PATEX["<b>emitted names — examples</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li><code>storage_op|storage_level=fs|fs_op=read|size_bytes=4096|idx=7</code></li><li><code>batch|dl_kind=getitem|idx=7|worker_id=0</code></li><li><code>batch|dl_kind=consume|seq=42</code></li><li><code>checkpoint|func=torch.save|total_bytes=1048576</code></li><li><code>chakra_global_metadata|backend_type=nvme|rank=0|world_size=2</code></li></ul>"]
        PAT -.-> PATEX
    end

    subgraph L3["Layer 3 · Kernel capture — Linux, ROOT REQUIRED"]
        direction LR
        SC["<b>marker_sidecar.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>sudo bpftrace subprocess</li><li>ready handshake before the workload starts</li></ul>"]
        MKPY["<b>marker.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>ctypes wrapper</li><li>no-ops if the .so is absent</li></ul>"]
        MKSO["<b>libstorage_overlay_marker.so</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>built from marker.c</li><li>no-op fns as uprobe attach points</li></ul>"]
        BT["<b>storage_overlay.bt</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>block · nvme · filemap</li><li>readahead · writeback</li></ul>"]
        SC --> BT
        MKPY --> MKSO
        MKSO -. "uprobe" .-> BT
    end

    subgraph L4["Layer 4 · Offline post-processing — torch-free, runs anywhere"]
        direction LR
        TM["<b>trace_merge.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>merge_per_rank — folds each worker trace into its rank’s main trace</li><li>kineto timing enrichment</li><li>produce / consume edge pairing</li></ul>"]
        FM["<b>fiemap.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>file to physical sector extents</li></ul>"]
        EI["<b>ebpf_inject.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>ATTRIBUTION — see diagram 4</li><li>folds kernel events into the host trace</li><li>parses and re-emits _patterns names</li></ul>"]
        TM --> EI
        FM --> EI
    end

    subgraph CH["mlcommons/chakra · storage_pass"]
        direction LR
        PC["<b>pytorch_converter.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>wiring</li><li>contract attr passthrough</li></ul>"]
        SP["<b>storage_pass.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>regions to typed STORAGE_NODE / DATA_OP_NODE</li><li>attaches StorageInfo</li></ul>"]
        PC --> SP
    end

    L1 == "names its regions with" ==> L2
    L1 -. "ctypes calls into the marker .so" .-> L3
    L1 -- "host_rank N .json — main" --> L4
    L1 -- "host_rank N _w W .json — per worker" --> L4
    L3 -- "ebpf_rank N .jsonl" --> L4
    L2 == "grammar parsed by" ==> CH
    L4 -- "merged host JSON" --> CH

    style L1 fill:#e8f4ea,stroke:#4a7a5a
    style L2 fill:#fff4d6,stroke:#a8862c,stroke-width:3px
    style PATEX fill:#fffaf0,stroke:#a8862c
    style L3 fill:#fde8e8,stroke:#a34d4d
    style L4 fill:#e8eef8,stroke:#4a6a9a
    style CH fill:#f0e8f8,stroke:#7a5a9a
```

---

## 2. End-to-end tracing workflow

From `enable_storage_capture()` to a typed `.et`. Rounded boxes are artifacts on disk; rectangles are steps.

```mermaid
flowchart TB
    A1["<b>1 · enable_storage_capture</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>workload calls enable_storage_capture(dataset_root)</li><li>installs the torch-side hooks</li><li>registers block devices + file inodes via so_scope_dev / so_scope_inode</li><li>_scope_devs_for walks /sys/dev/block/*/slaves for LVM and dm stacks</li></ul>"]

    B1["<b>2 · run_with_ebpf.sh</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>launches the sidecar FIRST, attached to the wrapper process pid</li><li>sched_process_fork picks up the training process as a child</li><li>polls --ready-file so probes are armed before the workload starts</li></ul>"]

    C1["<b>3 · training runs</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>main process + one observer per DataLoader worker</li><li>record_function regions on the host side</li><li>kernel events on the sidecar side</li></ul>"]
    HJ(["<b>host_rank N .json</b><br/><span style='white-space:nowrap'>main process</span>"])
    HW(["<b>host_rank N _w W .json</b><br/><span style='white-space:nowrap'>one per DataLoader worker</span>"])
    EJ(["<b>ebpf_rank N .jsonl</b><br/><span style='white-space:nowrap'>kernel eBPF events</span>"])

    D1["<b>4 · trace_merge.py</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>globs host_rank N _w*.json and rebases their node ids</li><li>each worker keeps its own thread root — a parallel spine</li><li>kineto timing enrichment</li><li>produce / consume edges paired</li></ul>"]

    E1["<b>5 · ebpf_inject.py — ATTRIBUTION</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>events sharing an agg key collapse into one supernode per 50 ms window (--window-ms)</li><li>block_io / nvme_io stay per-bio, never collapsed</li><li>supernodes nested under the read that caused them</li><li>see diagram 4</li></ul>"]
    E2["<b>FIEMAP runs here</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>on the host, while the dataset files still exist</li></ul>"]
    MJ(["<b>host_rank N _merged.json</b>"])

    F1["<b>6 · chakra_converter · storage_pass</b><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>promotes regions into typed STORAGE_NODE / DATA_OP_NODE</li><li>attaches StorageInfo</li></ul>"]
    ET(["<b>rank N .et</b>"])

    A1 --> B1 --> C1
    C1 --> HJ
    C1 --> HW
    C1 --> EJ
    HJ --> D1 --> E1
    HW --> D1
    EJ --> E1
    E2 -.-> E1
    E1 --> MJ --> F1 --> ET

    style E1 fill:#fde8e8,stroke:#a34d4d,stroke-width:3px
    style F1 fill:#f0e8f8,stroke:#7a5a9a,stroke-width:3px
    style E2 fill:#fffaf0,stroke:#a8862c
    style HJ fill:#eeeeee,stroke:#888888
    style HW fill:#eeeeee,stroke:#888888
    style EJ fill:#eeeeee,stroke:#888888
    style MJ fill:#eeeeee,stroke:#888888
    style ET fill:#eeeeee,stroke:#888888
```

---

## 3. How Python context reaches a kernel event

The mechanism by which block I/O can be attributed to a dataset sample. The marker library is a set of functions that do nothing — they exist purely as stable symbols for bpftrace to attach uprobes to. The `dev, sector` keying set up in step 7 is what survives the loss of thread context by step 10. Without it, every completion event would be unattributable.

```mermaid
sequenceDiagram
    autonumber
    participant W as Training thread
    participant M as marker.py<br/>ctypes
    participant SO as libstorage_overlay_marker.so<br/>no-op fns
    participant BPF as bpftrace probes
    participant K as Kernel block layer
    participant J as ebpf_rank N .jsonl

    Note over SO,BPF: uprobes attached at sidecar startup
    W->>M: getitem_enter idx=7, worker_id=0
    M->>SO: so_getitem_enter 7, 0
    SO-->>BPF: uprobe fires, args read from<br/>the calling convention
    BPF->>BPF: store idx, worker_id<br/>in per-tid map

    W->>K: read of a dataset file
    K->>BPF: block_rq_issue tracepoint
    BPF->>BPF: look up per-tid map,<br/>stamp idx + worker_id onto the bio,<br/>keyed by dev, sector
    BPF->>J: emit event

    Note over K,BPF: completion fires in IRQ context —<br/>no thread context left to read
    K->>BPF: block_rq_complete
    BPF->>BPF: resolve via the dev, sector key<br/>stored at issue time
    BPF->>J: emit event with idx + worker_id

    W->>M: getitem_exit idx=7, worker_id=0
    M->>SO: so_getitem_exit 7, 0
    SO-->>BPF: clear per-tid map entry
```

---

## 4. Attribution — deciding which graph node an event hangs under

**The problem:** a `bio` carries no `dataset_idx`. The page cache decouples `read()` from disk I/O. Completions fire in IRQ context and readahead runs on `kworker`. So every nesting edge below `read` has to be *reconstructed*. This flowchart is that reconstruction.

```mermaid
flowchart TB
    EV(["<span style='white-space:nowrap'><b>kernel event from the sidecar</b></span><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>dev · sector · bytes</li><li>maybe idx, worker_id</li></ul>"])
    Q1{"in SCOPE_DEVS?"}
    DROP["<span style='white-space:nowrap'><b>dropped</b></span><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>not dataset I/O</li></ul>"]

    Q2{"sector maps into a<br/>file's FIEMAP extent?"}
    M1["<span style='white-space:nowrap'><b>MECHANISM 1 · sector correlation</b></span><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>sector to file to page offset</li><li>then to the specific read that covered it</li></ul>"]
    R1(["<span style='white-space:nowrap'><b>nest under that exact fs_read</b></span><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>read to page to block to nvme</li></ul>"])

    Q3{"carries a marker<br/>idx, worker_id?"}
    M2["<span style='white-space:nowrap'><b>MECHANISM 2 · marker stamping</b></span><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>the uprobe recorded the active getitem on this tid</li></ul>"]
    Q4{"does that getitem have<br/>exactly one fs_read?"}
    R2a(["<span style='white-space:nowrap'><b>nest under that getitem's single fs_read</b></span>"])
    R2b(["<span style='white-space:nowrap'><b>nest under the getitem region itself</b></span><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>the getitem read several files — which one served this event is unknowable</li><li>stop at the region rather than guess</li></ul>"])

    M3["<span style='white-space:nowrap'><b>MECHANISM 3 · last resort</b></span>"]
    R3(["<span style='white-space:nowrap'><b>synthetic outside_getitem region</b></span><ul style='text-align:left;margin:0;padding-left:1.1em;white-space:nowrap'><li>under the thread root</li><li>counted as unlinked</li></ul>"])

    EV --> Q1
    Q1 -- no --> DROP
    Q1 -- yes --> Q2
    Q2 -- yes --> M1 --> R1
    Q2 -- no --> Q3
    Q3 -- yes --> M2 --> Q4
    Q4 -- yes --> R2a
    Q4 -- no --> R2b
    Q3 -- no --> M3 --> R3

    style M1 fill:#e8f4ea,stroke:#4a7a5a
    style M2 fill:#fff4d6,stroke:#a8862c
    style M3 fill:#fde8e8,stroke:#a34d4d
    style DROP fill:#eeeeee,stroke:#888888
```

---

## 5. What the resulting Chakra graph looks like

The shape the whole pipeline exists to produce.

```mermaid
flowchart TB
    ROOT["thread root"]
    GI0["DATA_OP_NODE<br/>data_op = DATALOADER<br/>batch dl_kind=getitem idx=0"]
    GI1["DATA_OP_NODE<br/>data_op = DATALOADER<br/>batch dl_kind=getitem idx=1"]
    OPEN["STORAGE_NODE · fs<br/>FS_OP_OPEN"]
    READ["STORAGE_NODE · fs<br/>FS_OP_READ<br/>bytes, offset, path"]
    PAGE["STORAGE_NODE · page<br/>PAGE_OP_ADD_TO_CACHE"]
    BLOCK["STORAGE_NODE · block<br/>BLOCK_EVENT_COMPLETE<br/>sector, ts_queue_ns"]
    NVME["STORAGE_NODE · device<br/>nvme_setup_cmd · slba"]
    CLOSE["STORAGE_NODE · fs<br/>FS_OP_CLOSE"]
    CKPT["DATA_OP_NODE<br/>data_op = CHECKPOINT"]
    CWRITE["STORAGE_NODE · fs<br/>FS_OP_WRITE"]
    OG["outside_getitem<br/>synthetic · unlinked readahead"]

    ROOT --> GI0 --> GI1
    ROOT --> OG
    GI0 --> OPEN --> READ
    READ --> PAGE --> BLOCK --> NVME
    NVME -.-> CLOSE
    ROOT --> CKPT --> CWRITE

    style GI0 fill:#e8e0f4,stroke:#7a5a9a
    style GI1 fill:#e8e0f4,stroke:#7a5a9a
    style CKPT fill:#e8e0f4,stroke:#7a5a9a
    style OPEN fill:#d6ecfa,stroke:#3a7aaa
    style READ fill:#d6ecfa,stroke:#3a7aaa
    style CLOSE fill:#d6ecfa,stroke:#3a7aaa
    style CWRITE fill:#d6ecfa,stroke:#3a7aaa
    style PAGE fill:#d6f4f4,stroke:#3a9a9a
    style BLOCK fill:#fae0c8,stroke:#c07a2a
    style NVME fill:#f4d6f0,stroke:#aa3a8a
    style OG fill:#eeeeee,stroke:#888888
```
